rfd/0163-vnet.md

```diff
@@ -0,0 +1,971 @@
+
+Once VNet starts, the VNet address of the app is copied to clipboard along with an in-app
+notification about the clipboard contents. Afterwards the button becomes just "Copy VNet address"
+when VNet is running. Copying the address copies `spec.public_addr` but with HTTP as the protocol.
+
+For TCP apps, the main "Connect" button is the equivalent of "Connect and copy VNet address". From
+the three dots menu, the user can select the sole option "Connect to local port". The address
+displayed on the card of a TCP app is the `spec.public_addr` of the app.
+
+Selecting an app from the search bar performs the same action as clicking on the main button in the
+app card.
+
+#### Managing VNet
+
+<img width="871" alt="VNet in Connections" src="assets/0163-vnet-in-connections.png">
+
+VNet has a persistent place in the UI in the connection list in top left. From there, the user can
+see VNet status and start or stop VNet, similar to how they can disconnect existing database, SSH
+and Kube connections.
+
+Clicking on the VNet item opens a dedicated VNet panel within the connection list. The panel
+displays the DNS zones that are currently being proxied. Any errors related to starting or stopping
+VNet are displayed there as well.
+
+The panel also shows recent connections made through VNet. A green dot means that a tunnel is
+currently open. A gray dot means the tunnel was opened but has since been closed. An error dot means
+that a connection for the given app could not be established.
+
+Clicking on one of the connections copies the app address to the clipboard, along with the full
+error message if any.
+
+<img width="871" alt="VNet panel" src="assets/0163-vnet-panel.png">
+
+#### Lifecycle
+
+The first version of VNet integration in Connect behaves similarly to Connect My Computer. After the
+user starts VNet and until they explicitly stop it, VNet starts automatically when launching the
+app. It prompts for a password for root privileges through osascript on each start.
+
+Unfortunately, [Touch ID prompts are reserved for Apple
+clients](https://forums.developer.apple.com/forums/thread/701118?answerId=705661022#705661022). We
+assume that typically the user is going to start VNet and then just leave it running, somewhat
+negating the downside of having to enter credentials on each start.
+
+When VNet is running and the user logs in to another cluster, the privileged process detects the new
+login and performs the necessary setup for the new root cluster.
+
+##### Expired certs
+
+As shown in the diagram in the [DNS queries and IP address
+assignment](#dns-queries-and-ip-address-assignment) section, there are two calls made to the
+cluster, one to get the details of an app and one to proxy a connection to the cluster. If any of
+those calls fails due to expired certs, Connect prompts the user to log in again. It's similar to
+how Connect prompts for relogin after the cert of a local db proxy expires.
+
+##### Considered alternatives
```

Comment:
Assuming for a moment that the IP aspect is not necessary (it seems like direct IP references are not highly supported due to the need to query DNS first anyways), another way to get this behavior could be to have the proxy support socks5.  Although this may also conflict with existing cooperate proxy needs.

Another option which also does not provide the IP space, but could provide similar behavior is requiring a wildcard DNS to the proxy server, so the proxy can handle the request transparently.  This option would make supporting https much simpler, but is not really a vnet, just a smarter proxy.

Reply:
I don't understand how socks5 or wildcard DNS to the proxy play in here to be honest

The problem we're trying to solve is how can clients connect to teleport apps, without having to worry about authentication, because nobody wants to configure client mTLS certs or run `tsh proxy app`. With socks5 AFAIK the client needs to be configured for it and set up authentication. And if all requests go to the proxy, I don't understand how clients authenticate themselves. I feels like we are talking about different things here maybe there is a misunderstanding

Reply:
I had neglected the goal of trying to ease the authentication burden.  Socks5 could potentially do this, but I agree a wildcard DNS can not.  Since socks5 is not likely a good alternative I will go ahead and resolve